logstreams/src/main/java/io/zeebe/distributedlog/restore/LogReplicator.java

```diff
@@ -0,0 +1,108 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.zeebe.distributedlog.restore;
+
+import io.atomix.cluster.MemberId;
+import io.zeebe.distributedlog.restore.impl.DefaultLogReplicationRequest;
+import io.zeebe.util.ZbLogger;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import org.slf4j.Logger;
+
+public class LogReplicator {
+  private final LogReplicationAppender appender;
+  private final LogReplicationClient client;
+  private final Executor executor;
+  private final Logger logger;
+
+  public LogReplicator(
+      LogReplicationAppender appender, LogReplicationClient client, Executor executor) {
+    this(appender, client, executor, new ZbLogger(LogReplicator.class));
+  }
+
+  public LogReplicator(
+      LogReplicationAppender appender,
+      LogReplicationClient client,
+      Executor executor,
+      Logger logger) {
+    this.appender = appender;
+    this.client = client;
+    this.executor = executor;
+    this.logger = logger;
+  }
+
+  public CompletableFuture<Long> replicate(MemberId server, long from, long to) {
+    final CompletableFuture<Long> result = new CompletableFuture<>();
+    replicateInternal(server, from, to, result);
+    return result;
+  }
+
+  private void replicateInternal(
+      MemberId server, long from, long to, CompletableFuture<Long> result) {
+    final LogReplicationRequest request = new DefaultLogReplicationRequest(from, to);
+    client
+        .replicate(server, request)
+        .whenCompleteAsync(
+            (r, e) -> {
+              if (e != null) {
+                logger.trace("Error replicating {} from {}", request, server, e);
+                result.completeExceptionally(e);
+              } else {
+                if (!r.isValid()) {
+                  logger.trace(
+                      "Received invalid response {} when requesting {} from {}",
+                      r,
+                      request,
+                      server);
+                  result.completeExceptionally(
+                      new InvalidLogReplicationResponse(server, request, r));
+                  return;
+                }
+
+                if (appendEvents(server, from, to, result, r)) {
+                  if (r.getToPosition() < to && r.hasMoreAvailable()) {
+                    replicateInternal(server, r.getToPosition(), to, result);
+                  } else {
+                    result.complete(r.getToPosition());
+                  }
+                }
+              }
+            },
+            executor);
+  }
+
+  private boolean appendEvents(
+      MemberId server,
+      long from,
+      long to,
+      CompletableFuture<Long> result,
+      LogReplicationResponse response) {
+    try {
+      final long appendResult =
+          appender.append(response.getToPosition(), response.getSerializedEvents());
```

Comment:
In every case, request.fromPosition is the position of the event that it already has. So it would also make sense, if the response always skips this event and starts with the next event.

Reply:
Yeah, I'd thought about delegating this logic to the appender, but we probably don't want to read the block for every single append command, since in the "normal" case there's no need to do so.